### PR TITLE
[NIAD-3131] Add ConfidentialityCode support for `Conditions`

### PR DESCRIPTION
**Pull request by martin-nhs**
## What

This PR introduces the capability to handle `confidentialityCodes` for` Conditions`. The implementation checks if a `Linkset` or `EhrComposition` contains a confidentiality code with the value `NOPAT`. If this code is present, meta security is added to the resource; otherwise, the code is ignored.

## Why

More details can be found on the [ticket](https://nhse-dsic.atlassian.net/browse/NIAD-3131).

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have updated the [Changelog](/CHANGELOG.md) with details of my change in the UNRELEASED section if this change will affect end users
- [ ] A corresponding change has been made to the [Mapping Documentation repository][mapping-docs]

[mapping-docs]: https://github.com/NHSDigital/patient-switching-adaptors-mapping-documentation

### CODE REVIEW

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java [hunk 1]
```diff
@@ -66,29 +81,32 @@ public class ConditionMapperTest {
...
     private static final String MEDICATION_STATEMENT_ORDER_ID = "ORDER_REF_ID";
     private static final String PRESCRIBE_ID = "PRESCRIBE_ID";
-    public static final String NAMED_STATEMENT_REF_ID = "NAMED_STATEMENT_REF_ID";
-    public static final String STATEMENT_REF_ID = "STATEMENT_REF_ID";
-    public static final String STATEMENT_REF_ID_1 = "STATEMENT_REF_ID_1";
-    public static final int EXPECTED_NUMBER_OF_EXTENSIONS = 4;
+    static final String NAMED_STATEMENT_REF_ID = "NAMED_STATEMENT_REF_ID";
+    static final String STATEMENT_REF_ID = "STATEMENT_REF_ID";
+    static final String STATEMENT_REF_ID_1 = "STATEMENT_REF_ID_1";
+    static final int EXPECTED_NUMBER_OF_EXTENSIONS = 4;
```
Comment (martin-nhs) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java:87]: These likely can be `private`.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java [hunk 2]
```diff
@@ -289,6 +307,72 @@ public void mapConditionWithSnomedCodeInCoding() {
...
+        when(confidentialityService.createMetaAndAddSecurityIfConfidentialityCodesPresent(
+            eq(META_PROFILE),
+            confidentialityCodeCaptor.capture(),
+            confidentialityCodeCaptor.capture()
+        )).thenReturn(MetaFactory.getMetaFor(META_WITH_SECURITY, META_PROFILE));
+
+        final List<Condition> conditions = conditionMapper
+            .mapResources(ehrExtract, patient, Collections.emptyList(), PRACTISE_CODE);
+
+        conditionMapper.addReferences(buildBundleWithNamedStatementObservation(), conditions, ehrExtract);
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java:329]: Is calling this method important, I think the production code we care about is within `mapResources` which is called on the line above. I suspect calling this `addReferences` has no effect on the outcome of the test, and is just a distraction.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java [hunk 3]
```diff
@@ -289,6 +307,72 @@ public void mapConditionWithSnomedCodeInCoding() {
...
 
+    @Test
+    void When_Condition_With_NopatConfidentialityCode_Expect_MetaFromConfidentialityServiceWithSecurity() {
+        final Meta metaWithSecurity = MetaFactory.getMetaFor(META_WITH_SECURITY, META_PROFILE);
+        final RCMRMT030101UKEhrExtract ehrExtract =
+            unmarshallEhrExtract("linkset_valid_nopat_confidentiality_code.xml");
+
+        when(dateTimeMapper.mapDateTime(
+            any(String.class)
+        )).thenReturn(EHR_EXTRACT_AVAILABILITY_DATETIME);
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java:318]: This is a bit of a smell, but not major.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java [hunk 4]
```diff
@@ -289,6 +307,72 @@ public void mapConditionWithSnomedCodeInCoding() {
...
+        when(confidentialityService.createMetaAndAddSecurityIfConfidentialityCodesPresent(
+            eq(META_PROFILE),
+            confidentialityCodeCaptor.capture(),
+            confidentialityCodeCaptor.capture()
+        )).thenReturn(MetaFactory.getMetaFor(META_WITH_SECURITY, META_PROFILE));
+
+        final List<Condition> conditions = conditionMapper
+            .mapResources(ehrExtract, patient, Collections.emptyList(), PRACTISE_CODE);
+
+        conditionMapper.addReferences(buildBundleWithNamedStatementObservation(), conditions, ehrExtract);
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java:362]: Same `addReferences` here.

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java [hunk 5]
```diff
@@ -367,8 +451,21 @@ private Bundle buildBundleWithStatementRefObservations() {
                 .setResource(new Observation().setId(STATEMENT_REF_ID_1)));
     }
 
+    private void assertConditionsMetaIsExpected(List<Condition> conditions, Meta expectedMeta) {
+        conditions.forEach(condition -> assertThat(condition.getMeta()).usingRecursiveComparison().isEqualTo(expectedMeta));
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java:455]: I'd change this to be something like...

```suggestion
        assertAll(conditions.map(condition -> 
                () -> assertThat(condition.getMeta()).usingRecursiveComparison().isEqualTo(expectedMeta)
        ));
```

So we get the benefit of the `assertAll`

File: gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java [hunk 6]
```diff
@@ -367,8 +451,21 @@ private Bundle buildBundleWithStatementRefObservations() {
                 .setResource(new Observation().setId(STATEMENT_REF_ID_1)));
     }
 
+    private void assertConditionsMetaIsExpected(List<Condition> conditions, Meta expectedMeta) {
```
Comment (adrianclay) [inline gp2gp-translator/src/test/java/uk/nhs/adaptors/pss/translator/mapper/ConditionMapperTest.java:454]: `assertAllConditionsHaveMeta` maybe?

**Review by adrianclay — APPROVED** [review]

**Pull request merged**